### PR TITLE
Fiona <-> OGR field mappings take 2 (refactor)

### PR DESCRIPTION
**Pull request by sgillies**
Follow up on #1358.

The project's original mappings predated OGR's sub-types (https://gdal.org/development/rfc/rfc50_ogr_field_subtype.html) and those were kinda bolted on. In this version we embrace them. We map `(fieldtype, subtype)` tuples to Fiona field types.

The extremely gross sections of `if/else` conditionals in the OGR and Fiona feature builders have been reduced to looking up a class in one of the maps and calling one of its methods. The increase in lines of code is largely explained by the new small classes and lack of normalization.

Original tests pass, save one, which had the wrong expected value.

Benchmarking: this script runs in 0.49 seconds with Fiona 1.9.6 and 0.44 seconds with the code in this PR. Measured by cProfile. So, a ~10% speedup for a 288 record dataset with features that have 60+ fields.

```python
import fiona

with fiona.open(
    "ne_10m_admin_0_countries_lakes/ne_10m_admin_0_countries_lakes.shp"
) as src:
    with fiona.open("/tmp/test_perf.shp", "w", **src.profile) as dst:
        dst.writerecords(src)
```

This PR is about making the code more readable and easier to maintain. A small speedup is a nice bonus.

When reviewing, use the "split" view, it's much easier to read for these particular changes.

### CODE REVIEW

File: fiona/ogrext.pyx [hunk 1]
```diff
@@ -215,7 +205,142 @@ class TZ(datetime.tzinfo):
...
+                val = string_list[string_list_index]
+                try:
+                    val = val.decode(encoding)
+                except UnicodeDecodeError:
+                    log.warning(
+                        "Failed to decode %s using %s codec", val, encoding
+                    )
+                vals.append(val)
+                string_list_index += 1
+        return vals
```
Comment (sgillies) [inline fiona/ogrext.pyx:343]: Big TODO: consider moving these methods to the fiona type classes?

File: fiona/ogrext.pyx [hunk 2]
```diff
@@ -541,6 +603,94 @@ def featureRT(feat, collection):
...
+cdef class BinaryFieldGetter(AbstractFieldGetter):
+    cdef object get(self, OGRFieldDefnH fdefn):
+        return "bytes"
+
+
+cdef class StringListFieldGetter(AbstractFieldGetter):
+    cdef object get(self, OGRFieldDefnH fdefn):
+        return "List[str]"
+
+
```
Comment (sgillies) [inline fiona/ogrext.pyx:693]: Big TODO: consider moving these methods to the fiona type classes?

File: fiona/schema.pyx [hunk 3]
```diff
@@ -70,13 +150,11 @@ def normalize_field_type(ftype):
...
         An element from FIELD_TYPES
     """
-    if ftype in FIELD_TYPES:
+    if ftype in FIELD_TYPES_NAMES:
         return ftype
-    elif ftype == 'bool':
-        return 'bool'
     elif ftype.startswith('int'):
         width = int((ftype.split(':')[1:] or ['0'])[0])
-        if GDAL_VERSION_NUM >= 2000000 and (width == 0 or width >= 10):
```
Comment (sgillies) [inline fiona/schema.pyx:79]: We require GDAL 3.4+

File: tests/test_collection.py [hunk 4]
```diff
@@ -535,6 +535,7 @@ def test_shape_driver_rounds_float_property_to_requested_digits_number(self):
             assert 12.2 == rf1.properties["property1"]
             assert 12.9 == rf2.properties["property1"]
 
+    @pytest.mark.xfail(reason="Corner case not supported by new feature builder implementation")
```
Comment (sgillies) [inline tests/test_collection.py:538]: Currently, I'm xfail-ing 4 tests that assert some weird behavior. For example, I'm not sure that Fiona should be ensuring that when you write `"3.14"` (a string) to an int field, you get a `3` (int) in your shapefile. This would require wrapping the setter's input value like `int(float(value))` and that's not negligible:

```
python3 -m timeit "int(3.14)"
10000000 loops, best of 5: 26.8 nsec per loop
python3 -m timeit "int(float('3.14'))"
5000000 loops, best of 5: 53.5 nsec per loop
```

File: tests/test_props.py [hunk 5]
```diff
@@ -24,9 +28,10 @@ def test_types():
     assert prop_type("str") == str
     assert isinstance(0, prop_type("int"))
     assert isinstance(0.0, prop_type("float"))
-    assert prop_type("date") == FionaDateType
+    assert prop_type("date") == str
```
Comment (sgillies) [inline tests/test_props.py:31]: `FionaDateType` no longer derives from `str`.

File: tests/test_schema.py [hunk 6]
```diff
@@ -200,15 +200,18 @@ def test_normalize_float():
...
 def generate_field_types():
     """
     Produce a unique set of field types in a consistent order.
 
     This ensures that tests are able to run in parallel.
     """
-    types = set(FIELD_TYPES)
-    types.remove(None)
-    return list(sorted(types)) + [None]
+    types = set(NAMED_FIELD_TYPES.keys())
```
Comment (sgillies) [inline tests/test_schema.py:213]: `FIELD_TYPES` remains, but internal usage now uses the more complete `NAMED_FIELD_TYPES` to get, for example `FionaBooleanType` by its name `"bool"`.

File: fiona/ogrext.pyx [hunk 7]
```diff
@@ -215,7 +205,351 @@ class TZ(datetime.tzinfo):
...
+        """Get the short Fiona field name corresponding to the OGR field definition."""
+        raise NotImplementedError
+
+    cdef object get(self, OGRFeatureH feature, int i, object kwds):
+        """Get the value of a feature's field."""
+        raise NotImplementedError
+
+    cdef set(self, OGRFeatureH feature, int i, object value, object kwds):
+        """Set the value of a feature's field."""
+        raise NotImplementedError
```
Comment (sgillies) [inline fiona/ogrext.pyx:227]: The classes that derive from this one replace the ugly if/def statements in the feature builders.

File: fiona/ogrext.pyx [hunk 8]
```diff
@@ -225,7 +559,37 @@ cdef class FeatureBuilder:
     argument is not destroyed.
     """
 
-    cdef build(self, void *feature, encoding='utf-8', bbox=False, driver=None, ignore_fields=None, ignore_geometry=False):
+    cdef object driver
+    cdef object property_getter_cache
```
Comment (sgillies) [inline fiona/ogrext.pyx:563]: I've added some state to the class so that we can reuse field getter and setters.

File: fiona/ogrext.pyx [hunk 9]
```diff
@@ -248,142 +612,65 @@ cdef class FeatureBuilder:
...
-                    props[key] = bool(OGR_F_GetFieldAsInteger(feature, i))
-                else:
-                    props[key] = OGR_F_GetFieldAsInteger(feature, i)
-
-            elif fieldtype is int:
-                if fieldsubtype == OFSTBoolean:
-                    props[key] = bool(OGR_F_GetFieldAsInteger64(feature, i))
+            else:
+                if fieldkey in self.property_getter_cache:
+                    getter = self.property_getter_cache[fieldkey]
```
Comment (sgillies) [inline fiona/ogrext.pyx:656]: Here we try to reuse an existing field getter.

File: fiona/ogrext.pyx [hunk 10]
```diff
@@ -248,142 +612,65 @@ cdef class FeatureBuilder:
...
-                            val = val.decode(encoding)
-                        except UnicodeDecodeError:
-                            log.warning(
-                                "Failed to decode %s using %s codec", val, encoding
-                            )
-                        props[key].append(val)
-                        string_list_index += 1
-            else:
-                props[key] = None
+                props[key] = getter.get(feature, i, {"encoding": encoding})
```
Comment (sgillies) [inline fiona/ogrext.pyx:669]: All that old crap replaced by this 😄

File: fiona/ogrext.pyx [hunk 11]
```diff
@@ -429,89 +739,32 @@ cdef class OGRFeatureBuilder:
...
+                            fieldkey
+                        )
+                        continue
+
+                # Special case: serialize dicts to assist OGR.
+                if isinstance(value, dict):
+                    value = json.dumps(value)
+
+                setter.set(cogr_feature, i, value, {"encoding": encoding})
+
```
Comment (sgillies) [inline fiona/ogrext.pyx:767]: I'm immoderately pleased about this.

File: fiona/ogrext.pyx [hunk 12]
```diff
@@ -1367,14 +1597,15 @@ cdef class WritingSession(Session):
...
+        cdef OGRSFDriverH cogr_driver
+        cdef OGRFeatureH cogr_feature
         cdef int features_in_transaction = 0
-        cdef void *cogr_layer = self.cogr_layer
+        cdef OGRLayerH cogr_layer = self.cogr_layer
 
         if cogr_layer == NULL:
             raise ValueError("Null layer")
 
+        cdef OGRFeatureBuilder feat_builder = OGRFeatureBuilder(driver=collection.driver)
```
Comment (sgillies) [inline fiona/ogrext.pyx:1608]: Here's an optimization that I've overlooked for a while: reusing the feature builder.

File: tests/test_collection.py (hunk 4 above)
Comment (sgillies) [reply tests/test_collection.py:538]: Oh, now I understand the difference between the new and old code. Previously, we dispatched to `OGR_F_SetFieldXXX()` based on the type of the value, not on the type of the field. I may try that.

File: fiona/ogrext.pyx [hunk 13]
```diff
@@ -689,18 +689,36 @@ cdef class OGRFeatureBuilder:
...
+        (OFTDate, OFSTNone, "str"): DateField,
+        (OFTTime, OFSTNone, "time"): TimeField,
+        (OFTTime, OFSTNone, "str"): TimeField,
+        (OFTDateTime, OFSTNone, "datetime"): DateTimeField,
+        (OFTDateTime, OFSTNone, "str"): DateTimeField,
+        (OFTBinary, OFSTNone, "bytes"): BinaryField,
+        (OFTBinary, OFSTNone, "bytearray"): BinaryField,
+        (OFTBinary, OFSTNone, "memoryview"): BinaryField,
+        (OFTStringList, OFSTNone, "list"): StringListField,
+        (OFTString, OFSTJSON, "dict"): JSONField,
```
Comment (sgillies) [inline fiona/ogrext.pyx:721]: Dispatching to a field setter based on field type and value type restores the implicit behavior in fiona version 1.9.x.

**Comment by sgillies** [comment]
@mwtoews @snorfalorpagus @perrygeo I'm tagging you for review of this refactor PR. Meanwhile I'm going to work on benchmarking.

**Comment by sgillies** [comment]
Last call for review!
